Putting install scripts into dockerfile

shimwell opened the pull request:
work in progress, not ready for review

## Description
As mentioned in project https://github.com/svalinn/DAGMC/projects/6 it has been proposed to move the install scripts into the dockerfile.

## Motivation and Context
This simplifies the passing of environments to cmake and other commands as they are all visible in one file and no longer creating in one script to be used in another. 

## Changes
What kind of change does this PR introduce? (Bug fix, feature, documentation update...)
refactored dockerfile to include more ARGs ENVs and install commands.

## Behavior
What is the current behavior? What is the new behavior?
old = dockerfile copies in sh scripts and runs them, new = dockerfile runs install commands directly

shimwell committed:
moved install scripts into Dockerfile

shimwell committed:
improved docs/comments

shimwell committed:
added PR 817

shimwell committed:
Merge branch 'develop' into moving_scripts_into_dockerfile

shimwell commented:
When building this new dockerfile with the defaults (moab 5.4.0) I am getting a Pymoab / PYTHONPATH error which is similar to the error I am seeing on PR #816 

```bash
 58%] Building CXX object test/CMakeFiles/intx_in_plane_test.dir/intx_in_plane_test.cpp.o
    Complete output from command /usr/bin/python3 -c "import setuptools, tokenize;__file__='/root/build_dir/moab/bld/pymoab/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" develop --no-deps --prefix=/root/build_dir/moab/bld/pymoab:
    running develop
    Checking .pth file support in /root/build_dir/moab/bld/pymoab/lib/python3.6/site-packages
    /usr/bin/python3 -E -c pass
    TEST FAILED: /root/build_dir/moab/bld/pymoab/lib/python3.6/site-packages does NOT support .pth files
    error: bad install directory or PYTHONPATH
    
    You are attempting to install a package to a directory that is not
    on PYTHONPATH and which Python does not read ".pth" files from.  The
    installation directory you specified (via --install-dir, --prefix, or
    the distutils default setting) was:
    
        /root/build_dir/moab/bld/pymoab/lib/python3.6/site-packages
    
    and your PYTHONPATH environment variable currently contains:
    
        ''
    
    Here are some of your options for correcting the problem:
    
    * You can choose a different installation directory, i.e., one that is
      on PYTHONPATH or supports .pth files
    
    * You can add the installation directory to the PYTHONPATH environment
      variable.  (It must then also be on PYTHONPATH whenever you run
      Python and want to use the package(s) you are installing.)
    
    * You can set up the installation directory to support ".pth" files by
      using one of the approaches described here:
    
      https://setuptools.readthedocs.io/en/latest/easy_install.html#custom-installation-locations
    
    
    Please make the appropriate changes for your system and try again.
    
    ----------------------------------------
  Can't roll back pymoab; was not uninstalled
Command "/usr/bin/python3 -c "import setuptools, tokenize;__file__='/root/build_dir/moab/bld/pymoab/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" develop --no-deps --prefix=/root/build_dir/moab/bld/pymoab" failed with error code 1 in /root/build_dir/moab/bld/pymoab/
pymoab/CMakeFiles/pymoab-local-install.dir/build.make:58: recipe for target 'pymoab/CMakeFiles/pymoab-local-install' failed
make[2]: *** [pymoab/CMakeFiles/pymoab-local-install] Error 1
CMakeFiles/Makefile2:1238: recipe for target 'pymoab/CMakeFiles/pymoab-local-install.dir/all' failed
make[1]: *** [pymoab/CMakeFiles/pymoab-local-install.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

shimwell committed:
Merge branch 'develop' into moving_scripts_into_dockerfile

gonuke commented on CI/Dockerfile:
```diff
@@ -34,63 +55,139 @@ RUN apt-get -y update; \
 
-#setting the COMPILER variable
-ARG COMPILER=gcc
-ENV COMPILER=${COMPILER}
```
Don't we still need to set this ENV?  Or at least need to do so for at least one of the packages that we build?

shimwell replied:
The scripts use to accept a ```COMPILER``` argument that was then used to set ```CC``` and CXX``` arguments using some logic. 

if COMPILER == clang then change cc to clang and cxx to clang++
if COMPILER == gcc then change cc to gcc and cxx to g++

I've allowed the user to set CC and CXX directly and then we don't need the COMPILER arg

gonuke commented on CI/Dockerfile:
```diff
@@ -34,63 +55,139 @@ RUN apt-get -y update; \
-RUN /root/etc/CI/docker/build_embree.sh
+ENV embree_install_dir=${install_dir}/embree
+
+RUN mkdir EMBREE && \
```
This would be a good opportunity to impose some consistency in how we use variables and directory structures across our different builds.

For example, we could define the following variables for each package, `pkg`:
* `pkg_build_dir`
* `pkg_install_dir`
And then always use those directories as we `mkdir`, `cd`, etc

shimwell replied:
yep that sounds good, these sorts of improvements become more obvious when we see all the different ways the install scripts opperated in one place

shimwell replied:
I've had a bit of a tiny up from the original PR that was just a copy paste of the install scripts. I think the build folders are now more consistent

gonuke commented on CI/Dockerfile:
```diff
@@ -34,63 +55,139 @@ RUN apt-get -y update; \
+        -DEMBREE_TASKING_SYSTEM=INTERNAL \
+        -DEMBREE_TUTORIALS=OFF \
+        -DEMBREE_TBB_ROOT=/usr && \
+    make -j2 && make -j2 install && \
```
```suggestion
    make -j${ci_jobs} && make -j${ci_jobs} install && \
```

gonuke commented on CI/Dockerfile:
```diff
@@ -34,63 +55,139 @@ RUN apt-get -y update; \
+    rm -rf $HOME/EMBREE/embree
+
 
 FROM external_deps AS hdf5
```
We might be able to leave the HDF5 variations out of this for now since I think all the OS versions we may use have updated?

gonuke commented on CI/Dockerfile:
```diff
@@ -34,63 +55,139 @@ RUN apt-get -y update; \
-        /root/etc/CI/docker/build_moab.sh; \
-    fi;
+RUN mkdir -p ${moab_build_dir}/bld
+RUN cd ${moab_build_dir} && \
```
Can we skip the static build now?

shimwell replied:
I've never quite understood why two builds were required for moab. I've commented it out for now but also happy to remove it if it is not needed anymore.

gonuke reviewed (changes requested):
Thanks @shimwell - this looks like a good step forward! 

I recognize that some of these suggestions represent changes of intent that we may want to leave for a separate PR & discussion.

shimwell committed:
[skip ci] compile cores review suggestion by @gonuke

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

shimwell committed:
[skip ci] removed compiler arg fully

shimwell committed:
more consistent dirs for build

shimwell committed:
ordering and intenting args

shimwell committed:
reduced layers

gonuke commented on CI/Dockerfile:
```diff
@@ -34,63 +57,143 @@ RUN apt-get -y update; \
+ENV embree_build_dir=${build_dir}/embree
+
+RUN mkdir -p ${hdf5_build_dir}/build && \
+    cd ${hdf5_build_dir} && \
```
I think these are the wrong directory variables now :)

shimwell replied:
sorry about that, now fixed

gonuke commented on CI/Dockerfile:
```diff
@@ -34,63 +57,143 @@ RUN apt-get -y update; \
+                    -DEMBREE_TBB_ROOT=/usr && \
+    make -j${ci_jobs} && \
+    make -j${ci_jobs} install && \
+    rm -rf $HOME/EMBREE/embree
```
Can we use the directory variables here as well?  Will need to `cd` to a safe place first rather than delete a directory we are still in

shimwell replied:
Yes certainly, good spot.

shimwell committed:
added missing hdf5 install dir env

gonuke reviewed (changes requested):
A few issues introduced with the embree directory variables

shimwell committed:
commented out moab static build

shimwell committed:
corrected embree dirs

shimwell committed:
using embree dir in rm -rf cmd

shimwell committed:
cd to safe place

gonuke commented on CI/Dockerfile:
```diff
@@ -34,63 +57,148 @@ RUN apt-get -y update; \
+RUN mkdir -p ${embree_build_dir}/build && \
+    cd ${embree_build_dir} && \
+    git clone -b ${EMBREE_BRANCH} https://github.com/embree/embree && \
+    mkdir build && \
```
No longer needed - done in #103

shimwell replied:
Sorry I am not quite sure which part of this embree clone/install was done in https://github.com/svalinn/DAGMC/pull/103

I have removed the 2nd ```mkdir build``` if that is what was needed

gonuke commented on CI/Dockerfile:
```diff
@@ -34,63 +57,148 @@ RUN apt-get -y update; \
-        /root/etc/CI/docker/build_moab.sh; \
-    fi;
+# Set the hdf5 install dir as this is also needed by moab compile
+ENV hdf5_install_dir=${install_dir}/hdf5
```
This is done in line 130 isn't it?

gonuke replied:
Raising this one again...

shimwell replied:
I thought we needed to declare ```ENV``` in each build stage they are used as they were not carried between build stages.

gonuke commented on CI/Dockerfile:
```diff
@@ -34,63 +57,148 @@ RUN apt-get -y update; \
+                  -DENABLE_FORTRAN=OFF \
+                  -DENABLE_PYMOAB=ON && \
+    # ensure that the installation directory for python packages exist
+    for d in $(echo $PYTHONPATH | tr : '\n'); do mkdir -p $d; done && \
```
Why do we need this?

gonuke commented on CI/Dockerfile:
```diff
@@ -34,63 +57,148 @@ RUN apt-get -y update; \
+ENV dagmc_build_dir=${build_dir}/DAGMC
+ENV dagmc_install_dir=${install_dir}/DAGMC
+
+ENV dagmc_build_dir_shared=${dagmc_build_dir}-shared
+ENV dagmc_install_dir_shared=${dagmc_install_dir}-shared
+
+ENV dagmc_build_dir_static=${dagmc_build_dir}-static
+ENV dagmc_install_dir_static=${dagmc_install_dir}-static
```
Not sure if we need all of these?

shimwell replied:
now removed

gonuke reviewed (changes requested):
I think we're getting closer....

shimwell commented on CI/Dockerfile:
```diff
@@ -34,63 +57,148 @@ RUN apt-get -y update; \
+RUN mkdir -p ${embree_build_dir}/build && \
+    cd ${embree_build_dir} && \
+    git clone -b ${EMBREE_BRANCH} https://github.com/embree/embree && \
+    mkdir build && \
```
looks like I had two build dirs as well

```suggestion
```

shimwell committed:
removed 2nd mkdir build in embree compile

shimwell commented:
A moab compile error has appeared when trying to build this docker file locally with moab 5.4.0. 

```
Building CXX object test/CMakeFiles/intx_in_plane_test.dir/intx_in_plane_test.cpp.o
    Complete output from command /usr/bin/python3 -c "import setuptools, tokenize;__file__='/root/build_dir/moab/build/pymoab/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" develop --no-deps --prefix=/root/build_dir/moab/build/pymoab:
    running develop
    Checking .pth file support in /root/build_dir/moab/build/pymoab/lib/python3.6/site-packages
    /usr/bin/python3 -E -c pass
    TEST FAILED: /root/build_dir/moab/build/pymoab/lib/python3.6/site-packages does NOT support .pth files
    error: bad install directory or PYTHONPATH
    
    You are attempting to install a package to a directory that is not
    on PYTHONPATH and which Python does not read ".pth" files from.  The
    installation directory you specified (via --install-dir, --prefix, or
    the distutils default setting) was:
    
        /root/build_dir/moab/build/pymoab/lib/python3.6/site-packages
    
    and your PYTHONPATH environment variable currently contains:
    
        ''
    
    Here are some of your options for correcting the problem:
    
    * You can choose a different installation directory, i.e., one that is
      on PYTHONPATH or supports .pth files
    
    * You can add the installation directory to the PYTHONPATH environment
      variable.  (It must then also be on PYTHONPATH whenever you run
      Python and want to use the package(s) you are installing.)
    
    * You can set up the installation directory to support ".pth" files by
      using one of the approaches described here:
    
      https://setuptools.readthedocs.io/en/latest/easy_install.html#custom-installation-locations
    
    
    Please make the appropriate changes for your system and try again.
    
    ----------------------------------------
  Can't roll back pymoab; was not uninstalled
```

bam241 commented on CI/Dockerfile:
```diff
@@ -34,63 +57,147 @@ RUN apt-get -y update; \
 
 FROM hdf5 AS moab
 
-ENV PYTHONPATH=/root/build_dir/moab/bld/pymoab/lib/python3.6/site-packages
```
I think the CI is failing because this line is not present anymore :)

shimwell replied:
Yes I am sure you are right about this @bam241 I just wanted to avoid putting in this ENV to get things working. Ideally the moab install would sort out this

gonuke replied:
Can't we prepend this to the `make` command (or `make install`) for MOAB below rather than embed it in the `ENV`?

shimwell commented on CI/Dockerfile:
```diff
@@ -34,63 +57,148 @@ RUN apt-get -y update; \
+    cd && \
+    rm -rf ${moab_build_dir}
+
+ENV dagmc_build_dir=${build_dir}/DAGMC
+ENV dagmc_install_dir=${install_dir}/DAGMC
+
+ENV dagmc_build_dir_shared=${dagmc_build_dir}-shared
+ENV dagmc_install_dir_shared=${dagmc_install_dir}-shared
+
+ENV dagmc_build_dir_static=${dagmc_build_dir}-static
+ENV dagmc_install_dir_static=${dagmc_install_dir}-static
```
```suggestion
```
This was in the original install scripts when they were moved into the docker file but it doesn't look like they are needed or used anywhere

shimwell replied:
Looks like these env are just used in the CI scripts. Seeing as we are moving those into the dockerfile and I can't see anywhere else they are used I'm going to go ahead and remove them.
```
grep -r dagmc_install_dir_static .
./DAGMC/CI/env.sh:export dagmc_install_dir_static=${dagmc_install_dir}-static
./DAGMC/CI/scripts/tests.sh:PATH=${dagmc_install_dir_static}/bin:$PATH make test
./DAGMC/CI/scripts/build_dagmc.sh:    local install_dir=${dagmc_install_dir_static}
```

shimwell commented on CI/Dockerfile:
```diff
@@ -34,63 +57,148 @@ RUN apt-get -y update; \
+                  -DBUILD_SHARED_LIBS=ON \
+                  -DENABLE_FORTRAN=OFF \
+                  -DENABLE_PYMOAB=ON && \
+    # ensure that the installation directory for python packages exist
+    for d in $(echo $PYTHONPATH | tr : '\n'); do mkdir -p $d; done && \
```
```suggestion
```
This was in the original install scripts and moved into the dockerfile. It makes all the dirs in the python path. This shouldn't be necessary, the installer should make it's own folders.

gonuke replied:
Agreed that we shouldn't need it - @bam241 is this your addition?

bam241 replied:
I don't think so, at least not one that I recall

shimwell replied:
Looks like it was added in this commit https://github.com/svalinn/DAGMC/commit/12f905384089da0484fec6da043cb2ae556eb8b2

gonuke replied:
can we try without and see what happens?

shimwell committed:
remove env for dag

shimwell committed:
Remove making python path folders

gonuke commented:
For some reason, these latest commits didn't seem to trigger rebuild of the images on your fork???

shimwell commented:
Sorry about that, I've just manually trigger it
https://github.com/shimwell/DAGMC/actions/runs/2939772739

gonuke committed:
prepend pythonpath as fix for MOAB build

gonuke committed:
keep previous pythonpath as well

gonuke committed:
whitespace change to changelog

gonuke committed:
correct pythonpath

gonuke committed:
need to mkdir for temp install

gonuke committed:
add pythonpath for install as well

gonuke committed:
comment out env source b/c it should all be in docker image

gonuke committed:
Add build and install dir to the environment

gonuke committed:
define build_dir from env

gonuke committed:
read build dir from the env

gonuke committed:
workflow changes to correct directory

shimwell committed:
Merge pull request #4 from gonuke/moving_scripts_into_dockerfile

prepend pythonpath as fix for MOAB build

shimwell commented:
I just noticed that the last commit here passed the CI. Is it worth catching up on this PR at the next DAGMC meeting

gonuke reviewed (changes requested):
Just one extra definition of the HDF5 path.  It's not necessarily wrong, but if it remains, it could become inconsistent and become a very hard bug to track down.

shimwell committed:
moved hdf5_install_dir to global ARGs

shimwell commented:
I think that removing either of the ENV would cause the ENV not to be found in that build stage. So I opted for adding the hdf5 install path as a global ARG instead as that can be a single string that is accessed by both build stages.

gonuke commented:
> I think that removing either of the ENV would cause the ENV not to be found in that build stage. So I opted for adding the hdf5 install path as a global ARG instead as that can be a single string that is accessed by both build stages.

Hmmmm... I think I need to do some independent testing of this on a minimum example.  My understanding was somewhat reversed.  A different, but important, question is which is used to perform substitution in a docker command.  Is the ENV available for substitution in a RUN command that is in the dockerfile, or does it need to be an ARG anyway???

shimwell commented:
> 

I think it is the ```ARG``` values that can be set with a ```docker run build-arg``` command.

This thread has a few minimal examples in that might be handy
https://github.com/moby/moby/issues/37345

gonuke commented:
> This thread has a few minimal examples in that might be handy
> [moby/moby#37345](https://github.com/moby/moby/issues/37345)

This example is a little different because both stages start from the fresh base image  - the second stage doesn't build on the first stage.

Try this:
```
from ubuntu:20.04 as stage1

ENV foo=test1
ARG bar=test2
RUN echo ${foo}
RUN echo ${bar}
ENV foo=test3

from stage1 as stage2
RUN echo ${foo}
RUN echo ${bar}
```

shimwell committed:
single ENV entry used in multiple stages

shimwell commented:
Thanks for that example. I see the ```ENV``` got to the stage2 and the ```ARG``` didn't. Sorry for my misunderstanding
![Screenshot from 2023-01-01 22-31-56](https://user-images.githubusercontent.com/8583900/210186419-cad396f9-df73-45e7-a019-c9ecf2c152a7.png)

gonuke reviewed (approved):
I think this is OK for now.  I see the need for some cleanup down the road (e.g. ENV vs ARG) but it may need us to complete this transition for our CI infrastructure and then do a final coordinated cleanup over many parts of the CI workflow.

Thanks for sticking with this @shimwell.

I would like to see evidence of this running one more time - it doesn't seem to be triggering on your actions. 🤷‍♂️

shimwell committed:
Merge branch 'svalinn:develop' into moving_scripts_into_dockerfile

shimwell commented:
Super. I synced the fork which appears to have triggered some ci and manually trigger some docker file building CI over here
https://github.com/shimwell/DAGMC/actions/runs/3823442736

Note to self. I noticed a few of our status badges on the main readme are red

shimwell commented:
manual triggered CI failed while trying to compile MOAB

```error: bad install directory or PYTHONPATH```

pshriwise commented:
@shimwell I noted that in our latest CI run after merging #845. I have a proposed fix submitted to MOAB, so hopefully that will be resolved soon.

I'd vote we go ahead and merge this if @gonuke doesn't have further comments. Once that PR goes into MOAB, I'll make sure to re-run our CI here and confirm it fixes our issue.

gonuke commented:
But merging this will cause all of our CI images to fail and leave us without any CI, I think

shimwell commented:
Thanks for making [this](https://bitbucket.org/fathomteam/moab/pull-requests/607) python path fix PR @pshriwise mentioned.

I guess it would take effect with the next release of MOAB, would it make sense to therefore wait for the next release of MOAB before merging this PR that changes the dockerfile

pshriwise commented:
@gonuke I guess I was assuming that this would only affect the `master` and `develop` builds of MOAB, but even those failing would cause the others to be canceled, as seen in this build https://github.com/svalinn/DAGMC/actions/runs/3715840076 (the one successful job is present because I manually restarted it).

At any rate, my thought was that merging this isn't going to change the current state of our CI for a merge to our `develop`, unfortunately, so I figured we may as well wrap it up if there aren't any other issues.

I've opened up #847 to revisit how we handle experimental builds and continue this thread of discussion.

gonuke commented:
It is possible that this won't change our CI status - but only because it will fail :).  We do stage the new CI images and test them before we commit to them as the official base image for testing, so it's likely that we would not disrupt our current CI test images.  However, I worry about building in some technical debt by merging this now.

pshriwise commented:
Either way is good w/ me 👍🏻

shimwell committed:
moab to 5.3.0

shimwell commented:
I have downgraded to moab 5.3.0 for consistency with other parts of the CI

I have manually trigger the docker image building https://github.com/shimwell/DAGMC/actions/runs/3847616064

gonuke commented:
Looks like a new cause of failure - you might need to move/copy the contents of the "install.sh" script into to bootstrap test job further down this `docker_publish.yml` file

shimwell committed:
[skip ci] added double-down install, started adding dagmc stage

shimwell commented:
ah yes, sorry there is more work to be done here. Before this PR is ready

I proposed to:

1. add a stage for DAGMC building to the dockerfile
2.  rework the ```test_img``` stage of docker_publish.yml so that it builds this new stage of the Dockerfile with the various argument combinations.
3. Add new variables such as FC, static_exe, build_mw_reg_tests which I can introduce as ARGs to the Dockerfile

I have started on this with the last commit

shimwell committed:
building and testing dagmc in docker

shimwell commented:
I have added stages for building and testing dagmc to the dockerfile
I have also reworked the dockr_publish.yml to build these stages.

The docker build ci didn't need manually trigger this time and can been seen over [here](https://github.com/shimwell/DAGMC/actions/runs/3855379407)

The current error while building the dockerfile locally with default ARGS is on the last step where I run a ``` cmake . -DDAGMC_ROOT=${dagmc_install_dir} ``` command

```bash
Step 80/80 : RUN cd ${dagmc_build_dir}/cmake/test_config &&     cmake . -DDAGMC_ROOT=${dagmc_install_dir} &&     make all test
 ---> Running in 1f41a2bf7a80
-- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is GNU 7.5.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- /root/opt/dagmc
-- Found DAGMC in DAGMC_CMAKE_CONFIG-NOTFOUND
CMake Error at FindDAGMC.cmake:20 (include):
  include could not find load file:

    DAGMC_CMAKE_CONFIG-NOTFOUND/DAGMCConfig.cmake
Call Stack (most recent call first):
  CMakeLists.txt:11 (find_package)
```

shimwell committed:
removed static from tag

shimwell committed:
corrected install dir typo

shimwell committed:
merge conflict solved

shimwell commented:
After adding the dagmc_test stage to the dockerfile it took me a while to get it building locally

Happy to say that it is now working locally with the extra testing and building of dagmc

Let us see if the docker build CI works over here https://github.com/shimwell/DAGMC/actions/runs/3913008543

shimwell committed:
updated actions

shimwell committed:
added docker buildx to step

gonuke commented:
This is failing now in a strange place - `make_watertight` testing??

shimwell commented:
That is a new fail I've not seen before. I was building the dockerfile locally with the defaults and it built successfully. I can try building that particular combination of build args locally and try to reproduce.

Won't be able to do this till late next week as I am on hotel WiFi for a while

shimwell commented:
this is failing on my local CI https://github.com/shimwell/DAGMC/actions/runs/3917829965

shimwell commented:
I am trying to build the specific dockerfile with the same build args that failed with this command

```bash
docker build -t dagmc --build-arg UBUNTU_VERSION=18.04 --build-arg COMPILER=clang --build-arg HDF5=1.10.4 --build-arg MOAB=master --build-arg static_exe=OFF -f CI/Dockerfile .
```

shimwell committed:
added path before make test cmd

shimwell commented:
reporting on that docker image build with the build args that match the failing test

It build successfully with and without the addition ```PATH=${dagmc_install_dir}/bin:$PATH``` before the ```make test```

gonuke commented:
I think we need to turn on some verbose reporting of tests to figure this out

shimwell committed:
Merge branch 'svalinn:develop' into moving_scripts_into_dockerfile

shimwell commented:
> verbose

I'm not quite sure how to make a verbose output from the ```make all test``` test command.

Also the docker file is building locally which is nice, it passes the tests

```
-- Installing: /root/opt/dagmc/tests/small_overlap.h5m
Removing intermediate container db51e2b9bba7
 ---> b6de7c8119cc
Step 78/84 : FROM dagmc as dagmc_test
 ---> b6de7c8119cc
Step 79/84 : ARG dagmc_build_dir
 ---> Running in f3009d86ebaa
Removing intermediate container f3009d86ebaa
 ---> 2a9292f709b7
Step 80/84 : ARG dagmc_install_dir
 ---> Running in a4b7e3f86f8d
Removing intermediate container a4b7e3f86f8d
 ---> d741ee9bc61d
Step 81/84 : ENV VERBOSE=1
 ---> Running in 03c59aa9a7ba
Removing intermediate container 03c59aa9a7ba
 ---> da1142b01bf5
Step 82/84 : RUN cd ${dagmc_build_dir}/build &&     PATH=${dagmc_install_dir}/bin:$PATH make test
 ---> Running in 2e28c8cef8df
Running tests...
/usr/bin/ctest --force-new-ctest-process 
Test project /root/build_dir/dagmc/build
      Start  1: dagmc_unit_tests
 1/25 Test  #1: dagmc_unit_tests ........................   Passed    0.72 sec
      Start  2: dagmc_pointinvol_test
 2/25 Test  #2: dagmc_pointinvol_test ...................   Passed    0.71 sec
      Start  3: dagmc_rayfire_test
 3/25 Test  #3: dagmc_rayfire_test ......................   Passed    0.32 sec
      Start  4: dagmc_simple_test
 4/25 Test  #4: dagmc_simple_test .......................   Passed    0.38 sec
      Start  5: dagmc_graveyard_test
 5/25 Test  #5: dagmc_graveyard_test ....................   Passed   11.79 sec
      Start  6: uwuw_unit_tests
 6/25 Test  #6: uwuw_unit_tests .........................   Passed    0.09 sec
      Start  7: uwuw_unit_tests_tally
 7/25 Test  #7: uwuw_unit_tests_tally ...................   Passed    0.02 sec
      Start  8: uwuw_unit_tests_preprocessor
 8/25 Test  #8: uwuw_unit_tests_preprocessor ............   Passed    1.00 sec
      Start  9: test_KDEKernel
 9/25 Test  #9: test_KDEKernel ..........................   Passed    0.01 sec
      Start 10: test_KDEMeshTally
10/25 Test #10: test_KDEMeshTally .......................   Passed    1.31 sec
      Start 11: test_KDENeighborhood
11/25 Test #11: test_KDENeighborhood ....................   Passed    1.04 sec
      Start 12: test_PolynomialKernel
12/25 Test #12: test_PolynomialKernel ...................   Passed    0.01 sec
      Start 13: test_Quadrature
13/25 Test #13: test_Quadrature .........................   Passed    0.01 sec
      Start 14: test_CellTally
14/25 Test #14: test_CellTally ..........................   Passed    0.01 sec
      Start 15: test_TallyEvent
15/25 Test #15: test_TallyEvent .........................   Passed    0.00 sec
      Start 16: test_TallyData
16/25 Test #16: test_TallyData ..........................   Passed    0.04 sec
      Start 17: test_Tally
17/25 Test #17: test_Tally ..............................   Passed    6.67 sec
      Start 18: test_TrackLengthMeshTally
18/25 Test #18: test_TrackLengthMeshTally ...............   Passed  247.76 sec
      Start 19: make_watertight_cylinder_tests
19/25 Test #19: make_watertight_cylinder_tests ..........   Passed    1.68 sec
      Start 20: make_watertight_cone_tests
20/25 Test #20: make_watertight_cone_tests ..............   Passed  101.21 sec
      Start 21: make_watertight_no_curve_sphere_tests
21/25 Test #21: make_watertight_no_curve_sphere_tests ...   Passed   10.54 sec
      Start 22: make_watertight_sphere_n_box_test
22/25 Test #22: make_watertight_sphere_n_box_test .......   Passed   10.96 sec
      Start 23: dagmcnp_unit_tests
23/25 Test #23: dagmcnp_unit_tests ......................   Passed    0.28 sec
      Start 24: dagsolid_unit_tests
24/25 Test #24: dagsolid_unit_tests .....................   Passed    0.70 sec
      Start 25: overlap_check_test
25/25 Test #25: overlap_check_test ......................   Passed   59.56 sec

100% tests passed, 0 tests failed out of 25

Total Test time (real) = 456.85 sec
Removing intermediate container 2e28c8cef8df
 ---> 02c30c3ddf18
Step 83/84 : RUN cd ${dagmc_build_dir} &&     git clean -dxf .
 ---> Running in fc7fea2252ab
Removing build/
Removing src/pyne/pyne.cpp
Removing src/pyne/pyne.h
Removing intermediate container fc7fea2252ab
 ---> a6d3443ed6e3
Step 84/84 : RUN cd ${dagmc_build_dir}/cmake/test_config &&     cmake . -DDAGMC_ROOT=${dagmc_install_dir} &&     make all test
 ---> Running in a718dfb750a8
Re-run cmake no build system arguments
-- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is GNU 7.5.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- /root/opt/dagmc
-- Found DAGMC in /root/opt/dagmc/lib/cmake/dagmc
-- Configuring DAGMC 3.2.2
-- Configuring done
-- Generating done
-- Build files have been written to: /root/build_dir/dagmc/cmake/test_config
/usr/bin/cmake -H/root/build_dir/dagmc/cmake/test_config -B/root/build_dir/dagmc/cmake/test_config --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /root/build_dir/dagmc/cmake/test_config/CMakeFiles /root/build_dir/dagmc/cmake/test_config/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/root/build_dir/dagmc/cmake/test_config'
make -f CMakeFiles/test_DAGMCConfig.dir/build.make CMakeFiles/test_DAGMCConfig.dir/depend
make[2]: Entering directory '/root/build_dir/dagmc/cmake/test_config'
cd /root/build_dir/dagmc/cmake/test_config && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /root/build_dir/dagmc/cmake/test_config /root/build_dir/dagmc/cmake/test_config /root/build_dir/dagmc/cmake/test_config /root/build_dir/dagmc/cmake/test_config /root/build_dir/dagmc/cmake/test_config/CMakeFiles/test_DAGMCConfig.dir/DependInfo.cmake --color=
Dependee "/root/build_dir/dagmc/cmake/test_config/CMakeFiles/test_DAGMCConfig.dir/DependInfo.cmake" is newer than depender "/root/build_dir/dagmc/cmake/test_config/CMakeFiles/test_DAGMCConfig.dir/depend.internal".
Dependee "/root/build_dir/dagmc/cmake/test_config/CMakeFiles/CMakeDirectoryInformation.cmake" is newer than depender "/root/build_dir/dagmc/cmake/test_config/CMakeFiles/test_DAGMCConfig.dir/depend.internal".
Scanning dependencies of target test_DAGMCConfig
make[2]: Leaving directory '/root/build_dir/dagmc/cmake/test_config'
make -f CMakeFiles/test_DAGMCConfig.dir/build.make CMakeFiles/test_DAGMCConfig.dir/build
make[2]: Entering directory '/root/build_dir/dagmc/cmake/test_config'
[ 50%] Building CXX object CMakeFiles/test_DAGMCConfig.dir/test_DAGMCConfig.cpp.o
/usr/bin/c++   -I/root/opt/dagmc/include -I/root/opt/moab/include -I/usr/include/eigen3  -std=gnu++1z -o CMakeFiles/test_DAGMCConfig.dir/test_DAGMCConfig.cpp.o -c /root/build_dir/dagmc/cmake/test_config/test_DAGMCConfig.cpp
[100%] Linking CXX executable test_DAGMCConfig
/usr/bin/cmake -E cmake_link_script CMakeFiles/test_DAGMCConfig.dir/link.txt --verbose=1
/usr/bin/c++    -rdynamic CMakeFiles/test_DAGMCConfig.dir/test_DAGMCConfig.cpp.o  -o test_DAGMCConfig  -L/root/opt/dagmc/lib  -L/root/opt/moab/lib -Wl,-rpath,/root/opt/dagmc/lib:/root/opt/moab/lib -lMOAB -ldagmc -luwuw -lpyne_dagmc -ldagtally -lmakeWatertight -ldagsolid 
make[2]: Leaving directory '/root/build_dir/dagmc/cmake/test_config'
[100%] Built target test_DAGMCConfig
make[1]: Leaving directory '/root/build_dir/dagmc/cmake/test_config'
/usr/bin/cmake -E cmake_progress_start /root/build_dir/dagmc/cmake/test_config/CMakeFiles 0
Running tests...
/usr/bin/ctest --force-new-ctest-process 
Test project /root/build_dir/dagmc/cmake/test_config
    Start 1: DAGMC-Configuration-Test
1/1 Test #1: DAGMC-Configuration-Test .........   Passed    0.02 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.02 sec
Removing intermediate container a718dfb750a8
 ---> b67409f26766
[Warning] One or more build-args [COMPILER HDF5 MOAB] were not consumed
Successfully built b67409f26766
Successfully tagged dagmc:latest
(openmc_dev) jshimwell@FL000093:~/DAGMC-my_fork$ 
```

gonuke commented:
Maybe this? https://stackoverflow.com/questions/5709914/using-cmake-how-do-i-get-verbose-output-from-ctest

shimwell committed:
added inline CTEST_OUTPUT_ON_FAILURE=1

shimwell commented:
Attempting to rerun the dockerfile building over [here](https://github.com/shimwell/DAGMC/actions/runs/4076762574) on my fork CI

gonuke commented:
As @bam241 suspected, this has something to do with the tests not finding the H5M files needed for regression testing.

gonuke commented:
I think this may be an ARG/ENV thing.  Your `dagmc_test` stage only refers `dagmc_build_dir` as `ARG` but they are defined in the previous stage only as `ENV`.  I know we discussed this before, but this seems like a fragile construct.

I also noticed that the `dagmc` build is repeating all the previous stages.  This is because your docker build command is not using the already built container as a cache.  Instead, you are using it as the container on which to run docker.  Therefore, docker does not know that you have already completed the previous commands in the `Dockerfile`.  I'm sure it's possible to get this right, but I think we need to think about this all more carefully.

shimwell committed:
added ci stage for dagmc

shimwell committed:
dagmc_dirs as ENV not ARGS

shimwell commented:
Thanks Paul, I have had a go at fixing those two points.

The docker image now uses ENV for dagmc_build_dir and and ENV for dagmc_install_dir
Also the CI now has a stage that targets ```dagmc``` and another stage that targets ```dagmc_test```
I have renamed some tags of the resulting dockerfiles so that we don't overwritten previous ones.

The latest docker build CI can be seen [here](https://github.com/shimwell/DAGMC/actions/runs/4104092639)

shimwell committed:
corrected tag and container names

shimwell commented:
I just spotted an error with the tag names I had so I've corrected that and trying again

latest build action is [here](https://github.com/shimwell/DAGMC/actions/runs/4112551304)

shimwell committed:
pusing dagmc image so it can be found by testing image

shimwell commented:
CI failed to find image, I suspect this is due to it not being pushed, latest action can be found [here](https://github.com/shimwell/DAGMC/actions/runs/4116415022)

shimwell committed:
combining two steps

shimwell committed:
making use of qemu, removed buildx

shimwell committed:
housebuilding only if build-test works

shimwell committed:
housebuilding only if build-test works

shimwell commented:
:pray: work this time. Latest action link is [here](https://github.com/shimwell/DAGMC/actions/runs/4127114284)

shimwell committed:
docker qemu usage in dag step same as other stages

shimwell committed:
not running dag intsall step in container

shimwell commented:
Got to the stage where it [fails dagmc tests on my CI fork actions](https://github.com/shimwell/DAGMC/actions/runs/4135524101/jobs/7152217151)

Anyone got any ideas on this output error, looks like it can't find some h5m files (iter_imprinted.h5m and bllite30matls.h5m) which I also can't find on the repo

```
#17 266.9 23/26 Test #23: make_watertight_regression_tests ........***Failed    0.01 sec
#17 266.9 [==========] Running 3 tests from 3 test cases.
#17 266.9 [----------] Global test environment set-up.
#17 266.9 [----------] 1 test from ITERMakeWatertightRegressionTest
#17 266.9 [ RUN      ] ITERMakeWatertightRegressionTest.ITERRegressionTest
#17 266.9 MOAB ERROR: --------------------- Error Message ------------------------------------
#17 266.9 MOAB ERROR: iter_imprinted.h5m: No such file or directory!
#17 266.9 MOAB ERROR: serial_load_file() line 518 in Core.cpp
#17 266.9 MOAB ERROR: load_file() line 467 in Core.cpp
#17 266.9 /root/build_dir/dagmc/src/make_watertight/tests/test_classes.cpp:27: Failure
#17 266.9 Expected equality of these values:
#17 266.9   result
#17 266.9     Which is: 7
#17 266.9   moab::MB_SUCCESS
#17 266.9     Which is: 0
#17 266.9 /root/build_dir/dagmc/src/make_watertight/tests/test_classes.cpp:45: Failure
#17 266.9 Expected equality of these values:
#17 266.9   result
#17 266.9     Which is: 6
#17 266.9   moab::MB_SUCCESS
#17 266.9     Which is: 0
#17 266.9 /root/build_dir/dagmc/src/make_watertight/tests/test_classes.cpp:51: Failure
#17 266.9 Expected: (0) != (num_meshsets), actual: 0 vs 0
#17 266.9 MOAB ERROR: --------------------- Error Message ------------------------------------
#17 266.9 MOAB ERROR: file set not found!
#17 266.9 MOAB ERROR: get_sealing_mesh_tags() line 1976 in Gen.cpp
#17 266.9 MOAB ERROR: --------------------- Error Message ------------------------------------
#17 266.9 MOAB ERROR: could not get the mesh tags!
#17 266.9 MOAB ERROR: make_mesh_watertight() line 1803 in MakeWatertight.cpp
#17 266.9 MOAB ERROR: --------------------- Error Message ------------------------------------
#17 266.9 MOAB ERROR: could not make the mesh watertight!
#17 266.9 MOAB ERROR: seal_and_check() line 355 in test_classes.cpp
#17 266.9 [  FAILED  ] ITERMakeWatertightRegressionTest.ITERRegressionTest (1 ms)
#17 266.9 [----------] 1 test from ITERMakeWatertightRegressionTest (1 ms total)
#17 266.9 
#17 266.9 [----------] 1 test from FNSFMakeWatertightRegressionTest
#17 266.9 [ RUN      ] FNSFMakeWatertightRegressionTest.FNSFRegressionTest
#17 266.9 MOAB ERROR: --------------------- Error Message ------------------------------------
#17 266.9 MOAB ERROR: model1_360_1.h5m: No such file or directory!
#17 266.9 MOAB ERROR: serial_load_file() line 518 in Core.cpp
#17 266.9 MOAB ERROR: load_file() line 467 in Core.cpp
#17 266.9 /root/build_dir/dagmc/src/make_watertight/tests/test_classes.cpp:27: Failure
#17 266.9 Expected equality of these values:
#17 266.9   result
#17 266.9     Which is: 7
#17 266.9   moab::MB_SUCCESS
#17 266.9     Which is: 0
#17 266.9 /root/build_dir/dagmc/src/make_watertight/tests/test_classes.cpp:45: Failure
#17 266.9 Expected equality of these values:
#17 266.9   result
#17 266.9     Which is: 6
#17 266.9   moab::MB_SUCCESS
#17 266.9     Which is: 0
#17 266.9 /root/build_dir/dagmc/src/make_watertight/tests/test_classes.cpp:51: Failure
#17 266.9 Expected: (0) != (num_meshsets), actual: 0 vs 0
#17 266.9 MOAB ERROR: --------------------- Error Message ------------------------------------
#17 266.9 MOAB ERROR: file set not found!
#17 266.9 MOAB ERROR: get_sealing_mesh_tags() line 1976 in Gen.cpp
#17 266.9 MOAB ERROR: --------------------- Error Message ------------------------------------
#17 266.9 MOAB ERROR: could not get the mesh tags!
#17 266.9 MOAB ERROR: make_mesh_watertight() line 1803 in MakeWatertight.cpp
#17 266.9 MOAB ERROR: --------------------- Error Message ------------------------------------
#17 266.9 MOAB ERROR: could not make the mesh watertight!
#17 266.9 MOAB ERROR: seal_and_check() line 355 in test_classes.cpp
#17 266.9 [  FAILED  ] FNSFMakeWatertightRegressionTest.FNSFRegressionTest (0 ms)
#17 266.9 [----------] 1 test from FNSFMakeWatertightRegressionTest (0 ms total)
#17 266.9 
#17 266.9 [----------] 1 test from BLiteMakeWatertightRegressionTest
#17 266.9 [ RUN      ] BLiteMakeWatertightRegressionTest.BLiteRegressionTest
#17 266.9 MOAB ERROR: --------------------- Error Message ------------------------------------
#17 266.9 MOAB ERROR: bllite30matls.h5m: No such file or directory!
#17 266.9 MOAB ERROR: serial_load_file() line 518 in Core.cpp
#17 266.9 MOAB ERROR: load_file() line 467 in Core.cpp
#17 266.9 /root/build_dir/dagmc/src/make_watertight/tests/test_classes.cpp:27: Failure
#17 266.9 Expected equality of these values:
#17 266.9   result
#17 266.9     Which is: 7
#17 266.9   moab::MB_SUCCESS
#17 266.9     Which is: 0
#17 266.9 /root/build_dir/dagmc/src/make_watertight/tests/test_classes.cpp:45: Failure
#17 266.9 Expected equality of these values:
#17 266.9   result
#17 266.9     Which is: 6
#17 266.9   moab::MB_SUCCESS
#17 266.9     Which is: 0
#17 266.9 /root/build_dir/dagmc/src/make_watertight/tests/test_classes.cpp:51: Failure
#17 266.9 Expected: (0) != (num_meshsets), actual: 0 vs 0
#17 266.9 MOAB ERROR: --------------------- Error Message ------------------------------------
#17 266.9 MOAB ERROR: file set not found!
#17 266.9 MOAB ERROR: get_sealing_mesh_tags() line 1976 in Gen.cpp
#17 266.9 MOAB ERROR: --------------------- Error Message ------------------------------------
#17 266.9 MOAB ERROR: could not get the mesh tags!
#17 266.9 MOAB ERROR: make_mesh_watertight() line 1803 in MakeWatertight.cpp
#17 266.9 MOAB ERROR: --------------------- Error Message ------------------------------------
#17 266.9 MOAB ERROR: could not make the mesh watertight!
#17 266.9 MOAB ERROR: seal_and_check() line 355 in test_classes.cpp
#17 266.9 [  FAILED  ] BLiteMakeWatertightRegressionTest.BLiteRegressionTest (1 ms)
#17 266.9 [----------] 1 test from BLiteMakeWatertightRegressionTest (1 ms total)
#17 266.9 
#17 266.9 [----------] Global test environment tear-down
#17 266.9 [==========] 3 tests from 3 test cases ran. (2 ms total)
#17 266.9 [  PASSED  ] 0 tests.
#17 266.9 [  FAILED  ] 3 tests, listed below:
#17 266.9 [  FAILED  ] ITERMakeWatertightRegressionTest.ITERRegressionTest
#17 266.9 [  FAILED  ] FNSFMakeWatertightRegressionTest.FNSFRegressionTest
#17 266.9 [  FAILED  ] BLiteMakeWatertightRegressionTest.BLiteRegressionTest
#17 266.9 
#17 266.9  3 FAILED TESTS
#17 266.9 
```

bam241 commented:
maybe they are created by the tests.

But were the `tests` are running and where the `tests` used to be run are not the same location, so the files are not created where the tests are looking for them ?

bam241 commented:
or do the `test` have permission to write new file in those location ?

gonuke commented:
I think I found it!  There is a whole logic in the current system about whether it is a pull request or not (then a merge) that determines whether or not to run this regression test.  It should only run on merge.  Simultaneously, if it is not a pull request, there are lines to wget the missing files!!

shimwell commented:
Shall we simplify that logic so that it always downloads the model and runs the test

I can see this wget command in the [install.sh](https://github.com/svalinn/DAGMC/blob/af2a4ce0d9b2b7561c6ca9049d8e37fde61cdf7e/CI/scripts/install.sh#L17) script

```bash
  wget ${MW_REG_TEST_MODELS_URL} -O mw_reg_test_files.tar.gz -o wget.out
```

I did a quick grep for the ```MW_REG_TEST_MODELS_URL``` string but I can't see where it is set.

makeclean commented:
${docker-env} file?

makeclean commented:
All these bash scripts should have `set -u` in them for safety for undefined variables and '-e' if we want to exit on errors.

gonuke commented:
This PR is aiming to get rid of the scripts!

gonuke commented:
> Shall we simplify that logic so that it always downloads the model and runs the test
> 
> 
> 
> I can see this wget command in the [install.sh](https://github.com/svalinn/DAGMC/blob/af2a4ce0d9b2b7561c6ca9049d8e37fde61cdf7e/CI/scripts/install.sh#L17) script
> 
> 
> 
> ```bash
> 
>   wget ${MW_REG_TEST_MODELS_URL} -O mw_reg_test_files.tar.gz -o wget.out
> 
> ```
> 
> 
> 
> I did a quick grep for the ```MW_REG_TEST_MODELS_URL``` string but I can't see where it is set.

Hmmmm.... it appears to have been lost in CI transitions...

shimwell commented on CI/Dockerfile:
```diff
@@ -33,64 +61,208 @@ RUN apt-get -y update; \
+                      -DBUILD_GEANT4=ON \
+                      -DGEANT4_DIR=${geant4_install_dir} \
+                      -DBUILD_CI_TESTS=ON \
+                      -DBUILD_MW_REG_TESTS=${build_mw_reg_tests} \
```
this line could be changed to avoid running these tests.

The tests download a tar.gz file from a url that we can't find anymore

```bash
 -DBUILD_MW_REG_TESTS=OFF \
```

shimwell replied:
Discussed this at the DAGMC meeting yesterday and this is the preferred route, I've opened issue #858 to track the missing tar.gz file with the models in it.

shimwell committed:
disable reg tests as models missing

shimwell commented:
Latest attempt to get the dockerfile building is running on my forks actions over [here](https://github.com/shimwell/DAGMC/actions/runs/4203792529)

shimwell commented:
Action on my fork is passing 🎉

shimwell commented:
If this one gets merged can I make a follow up one that removes the old scripts or is that a step too far

gonuke commented:
Sorry @shimwell , but I'm just noticing that none of the incremental stages are using any of the prior dockerfiles at all.  You are building from scratch at each stage (and thus wasting a lot of CI resources :) ). Alas, I think we lost this without noticing in #813

gonuke commented:
Prior to #813, we had different dockerfiles for each stage, and each of those files used a different container that had already been pushed to GHCR.  With a single dockerfile, we have stages that build on each other.  At each step, we aim to build a particular stage, but with no caching, the CI runner has no idea that we've already built that stage.  As a result, every step is rebuilding GEANT, HDF5, etc... including the step where you test the DAGMC install.

gonuke commented:
The right way to fix this is probably introduce all the right caching, etc, but perhaps the multistage-docker-build-action is what we need to simplify this.

shimwell commented:
Good spot yes, I can start implementing the multistage-docker-build-action on this open PR

gonuke commented:
I think this has been replaced by #863 - do you want to close this one @shimwell ?

shimwell commented:
Closing in favour of #863
